### PR TITLE
NEW: First pass at cli for cache

### PR DESCRIPTION
**Pull request by Oddant1**
Start adding cli support for the cache so it can enter wider use

### CODE REVIEW

**Comment by Oddant1** [comment]
Regarding [b3a9f8b](https://github.com/qiime2/q2cli/pull/277/commits/b3a9f8b8aca0040c4a110fc5aec1452202a70155), this, unfortunately, loses us the pretty print when the path ends up not existing. If we want it back, we could break our API here and resolve a path within the cache's data directory to hand into this function. This would break our API, but it would be us breaking it internally so *shrug*. Not sure if we want to play with that though.

**Comment by ebolyen** [comment]
@Oddant1, what if we split the difference, if a path contains a colon, we presume it's probably a cache, and we attempt to resolve it as such. If it doesn't resolve, then we treat is at a file (on the off-chance the user hates things that make sense), and if it's not a file but did have the colon, then we raise an error about the key not existing in the cache (which is probably more true than it being a missing file).

Then, if something never had the colon, it continues to produce the file-doesn't-exist error.

(If I'm understanding the current situation correctly)

**Comment by Oddant1** [comment]
@ebolyen, the reason we lost the pretty print is that I got rid of the need for the file to exist in the click.Path argument because if they passed in a cache:key then a file with that path will. . . hopefully not exist. I can still do the logic the way you described, but we won't get the click pretty print from the file not existing because that processing is happening after inspect_metadata is called. There doesn't seem to be a good place to apply that logic before calling inspect_metadata

**Comment by Oddant1** [comment]
Blocked by qiime2/qiime2#645

**Comment by Oddant1** [comment]
@ebolyen, it would be really nice if we could unify how we load inputs and metadata a bit more because there is a whole lot of overlap, but they don't quite use the same machinery. You'll see what I mean if you look at load_metadata and _convert_input. May be out of the scope of this PR, but I already refactored a good chunk of this.

**Comment by Oddant1** [comment]
@ebolyen are you interested in having cd278e1 in this PR? If not, I can hold off on implementing it until this PR is merged. If you're fine with it, I can write up a unit test or two (I manually tested it by messing with my plugin manager)

**Comment by ebolyen** [comment]
Thanks @Oddant1! I think the commit looks great. I'll pull this all down soon and play around.

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:95]: An easier way to do this and the split is:
```python
cache_path, key = fp.rsplit(':', 1)
```
Which will split from the right, at most once. The unpacking should also always succeed as we have proved there is at least 1 colon in the path.

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:98]: We should probably make `Cache.is_cache` indifferent to the input type. Accepting either `str` or `Path`

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:111]: Is this branch reachable given the specific `FileNotFoundError` filter?

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:368]: This class name will leak to the user, which probably isn't ideal.

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:385]: I would create a helper using the aforementioned `rsplit` method to factor this shared bit out.

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:347]: We should migrate this to `Result.load()` it would be useful generally.

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:339]: A more idiomatic way to do this is to initialize `artifact = None` before your first `if ':' in fp:`, then you can make this line:

```python
if artifact is None:
```

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:374]: Maybe this should be `try_as_cache_input`?

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:332]: I think there may be one too many try's here, or at least I'm struggling to follow what's expected to happen in different situations.

I think this may need some refactoring, but I'm not really sure how to go about it either. It seems like there must be phases we can distill?

File: q2cli/util.py
Comment (ebolyen) [inline q2cli/util.py:299]: This kind of circles around the original problem. Perhaps we refactor `get_input` to return more than just the result? And then call-sites can error as they need/want to?

Exceptions are technically something you can pass around as any other object.

**Review by ebolyen — CHANGES_REQUESTED** [review]
Thanks @Oddant1! I think all of the changes to the q2cli guts are great, but these utilities are still a bit mind-bending. It's probably worth trying to rewrite them from scratch, I think they are just kind of tired after so much use-case accretion.

File: q2cli/util.py
Comment (Oddant1) [reply q2cli/util.py:98]: Turns out I already am doing that. I already caste it to a pathlib.Path at the beginning of is_cache.

File: q2cli/util.py
Comment (Oddant1) [reply q2cli/util.py:111]: Probably not. I wanted to be certain that if some other error did somehow get raised like that we didn't just squash it. Should I just remove the else?

File: q2cli/util.py
Comment (Oddant1) [reply q2cli/util.py:331]: The try here is handling all the errors with loading an Artifact that we understand. If we get an issue that we don't understand we raise a ControlFlowException that can be handled by the call site. The reason the call sites also have try's is so _convert_input can pretty print the failure if there is an exception and so load_metadata can try to load the file as metadata instead of as an artifact if the error it got was a ControlFlowException (any error loading the artifact that we don't understand).

File: q2cli/util.py
Comment (Oddant1) [reply q2cli/util.py:348]: In Result.load or deeper down in mount or something where we actually open things?

File: q2cli/util.py
Comment (Oddant1) [reply q2cli/util.py:373]: In theory, it could but it won't with either of the current call sites. I'm not sure I understand why it's a problem if it leaks anyway.

File: q2cli/util.py
Comment (Oddant1) [reply q2cli/util.py:298]: What do we gain from doing that though? Maybe I've just looked at this too much, so it makes too much sense to me now, but in get_input we raise a ControlFlowException if something we didn't anticipate happens, and we raise more informative errors if we do understand what happened. Why would we pass them around instead of just raising them and letting the call site handle them? What do we gain by returning them or something instead?

**Review by ebolyen — APPROVED** [review]
Ok, I've thrown the kitchen sink at this PR by this point. Everything seems to be working!

**Pull request merged**
